src/core/filters/meta/bpf/include/meta_filter.h

```diff
@@ -0,0 +1,230 @@
+	struct {
+		u8 type;
+		u8 nmemb;
+		u16 offt;
+	} l;
+	struct {
+		union {
+			u8 bin[META_TARGET_MAX];
+			u64 num;
+		} u;
+		u8 sz;
+		u8 cmp;
+	} t;
+};
+
+/* Probe configuration; the key is the target symbol address */
+struct {
+	__uint(type, BPF_MAP_TYPE_ARRAY);
+	__uint(max_entries, META_OPS_MAX);
+	__type(key, u32);
+	__type(value, union retis_meta_op);
+} filter_meta_map SEC(".maps");
+
+struct retis_meta_ctx {
+	/* base address to read from. */
+	void *base;
+	/* relative to base. */
+	u16 offset;
+	/* type of leaf value. */
+	u8 type;
+	/* optional */
+	u8 nmemb;
+	/* Target Info */
+	/* actual data. */
+	void *data;
+	/* size of data (optional). */
+	u8 sz;
+	/* operation. */
+	u8 cmp;
+};
+
+#define PTR_BIT 1 << 6
+#define SIGN_BIT 1 << 7
+
+const volatile u32 nmeta = 0;
+
+static __noinline long meta_process_ops(struct retis_meta_ctx *ctx)
+{
+	union retis_meta_op *val;
+	u64 ptr;
+	u32 k;
+	int i;
+
+	for (i = 0, k = 0; k < nmeta && k < META_OPS_MAX; i++, k++) {
+		val = bpf_map_lookup_elem(&filter_meta_map, &i);
+		if (!val)
+			return -1;
+
+		/* process target */
+		if (k == 0) {
+			ctx->data = &val->t.u;
+			ctx->cmp = val->t.cmp;
+			ctx->sz = val->t.sz;
+			continue;
+		}
+
+		/* Load Pointer */
+		if (val->l.type == PTR_BIT) {
+			if (bpf_probe_read_kernel(&ptr, sizeof(void *),
+						  (char *)ctx->base + (val->l.offt)) != 0)
+				return -1;
+
+			ctx->base = (void *)ptr;
+			continue;
+		}
+
+		/* Non intermediate */
+		ctx->offset = val->l.offt;
+		ctx->type = val->l.type;
+		ctx->nmemb = val->l.nmemb;
+	}
+
+	return 0;
+}
+
+static __always_inline
+bool cmp_num(u64 operand1, u64 operand2, bool sign, u8 cmp_type)
+{
+	switch (cmp_type) {
+	case RETIS_EQ:
+		return (operand1 == operand2);
+	case RETIS_NEQ:
+		return (operand1 != operand2);
+	case RETIS_GT:
+		return sign ? ((s64)operand1 > (s64)operand2) : ((u64)operand1 > (u64)operand2);
+	case RETIS_LT:
+		return sign ? ((s64)operand1 < (s64)operand2) : ((u64)operand1 < (u64)operand2);
+	case RETIS_GE:
+		return sign ? ((s64)operand1 >= (s64)operand2) : ((u64)operand1 >= (u64)operand2);
+	case RETIS_LE:
+		return sign ? ((s64)operand1 <= (s64)operand2) : ((u64)operand1 <= (u64)operand2);
+	default:
+		/* FIXME: report error. */
+		break;
+	}
+
+	return false;
+}
+
+static __always_inline
+unsigned int filter_bytes(struct retis_meta_ctx *ctx)
+{
+	char val[META_TARGET_MAX];
+	u32 sz;
+
+	/* if it is an array of chars use its size. Alternatively, use
+	 * the target size (probe read could fail in this case).
+	 * Note: for some reason the one-liner version of this fails to
+	 * generate code accepted by the verifier. Broken in two lines
+	 * to workaround that issue
+	 */
+	sz = ctx->nmemb ? : ctx->sz;
+	sz = MIN(sz, sizeof(val));
+
+	if (bpf_probe_read_kernel_str(val, sz, (char *)ctx->base + ctx->offset) < 0)
+		return 0;
+
+	const char *sp1 = ctx->data, *sp2 = val;
+	while (sz-- > 0 && sz < META_OPS_MAX && *sp1) {
+		if (!*sp1 != !*sp2)
+			break;
```

Comment:
sgtm.
This was initially implemented as a standalone `strncmp()` but then got open-coded here (mostly because of the need for the upper bound that defeated a bit its purpose).
Will change this part in the next iteration.

Reply:
I slightly changed this to:
```
	while (sz-- > 0 && sz < META_OPS_MAX &&
	       *sp1 && *sp2 && !(*sp1 - *sp2))
		sp1++, sp2++;

	return !(*sp1 - *sp2);
```